### PR TITLE
planner: eliminate empty projection below aggregation

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tics/issues/3389

Problem Summary: Empty projection causes issues when it's an MPP plan.

### What is changed and how it works?

Eliminate empty projections when it's below aggregation.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by time-and-fate** [comment]
The failed test case is irrelevant to this PR.

**Comment by time-and-fate** [comment]
Some execution plans in `TestMPPJoin` are changed from `HashAgg` to `StreamAgg`. This is caused by cost change since a `LogicalProjection` is eliminated.
The tables in this test case are small (less than 10 rows), so it is easy to be affected by small changes in the plan.
So I think this change is acceptable.

**Review by Reminiscent — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Review by LittleFall — APPROVED** [review]
LGTM

File: planner/core/testdata/integration_serial_suite_out.json [hunk 1]
```diff
@@ -251,59 +251,56 @@
...
-          "    └─HashAgg 1.00 batchCop[tiflash]  funcs:count(1)->Column#12",
-          "      └─HashJoin 8.00 batchCop[tiflash]  inner join, equal:[eq(test.d1_t.d1_k, test.fact_t.d1_k)]",
-          "        ├─ExchangeReceiver(Build) 2.00 batchCop[tiflash]  ",
-          "        │ └─ExchangeSender 2.00 batchCop[tiflash]  ExchangeType: Broadcast",
-          "        │   └─Selection 2.00 batchCop[tiflash]  not(isnull(test.d1_t.d1_k))",
-          "        │     └─TableFullScan 2.00 batchCop[tiflash] table:d1_t keep order:false",
-          "        └─Selection(Probe) 8.00 batchCop[tiflash]  not(isnull(test.fact_t.d1_k))",
-          "          └─TableFullScan 8.00 batchCop[tiflash] table:fact_t keep order:false"
+          "StreamAgg 1.00 root  funcs:count(1)->Column#11",
+          "└─TableReader 8.00 root  data:ExchangeSender",
```
Comment (fzhedu) [inline planner/core/testdata/integration_serial_suite_out.json:255]: why this PR change the agg implementation from two-phase agg -> stream agg?

File: planner/core/rule_column_pruning.go [hunk 2]
```diff
@@ -149,7 +149,21 @@ func (la *LogicalAggregation) PruneColumns(parentUsedCols []*expression.Column)
...
+	// This kind of Projection would cause some bugs for MPP plan and is safe to be removed.
+	// This kind of Projection should be removed in Projection Elimination, but currently PrunColumnsAgain is
+	// the last rule. So we specially handle this case here.
+	if childProjection, isProjection := child.(*LogicalProjection); isProjection {
+		if len(childProjection.Exprs) == 0 && childProjection.Schema().Len() == 0 {
+			childOfChild := childProjection.children[0]
+			la.SetChildren(childOfChild)
+		}
+	}
+	return nil
```
Comment (fzhedu) [inline planner/core/rule_column_pruning.go:166]: why not put these into the postOptimize() which eliminates projection at last?

Comment (fzhedu) [reply planner/core/rule_column_pruning.go:166]:
```
func postOptimize(sctx sessionctx.Context, plan PhysicalPlan) PhysicalPlan {
	plan = eliminatePhysicalProjection(plan)
...
```

File: planner/core/testdata/integration_serial_suite_out.json (hunk 1 above)
Comment (time-and-fate) [reply planner/core/testdata/integration_serial_suite_out.json:255]: I explained the reason in this comment.

https://github.com/pingcap/tidb/pull/29606#issuecomment-964054884

File: planner/core/rule_column_pruning.go (hunk 2 above)
Comment (time-and-fate) [reply planner/core/rule_column_pruning.go:166]: According to https://github.com/pingcap/tics/issues/3389#issuecomment-963852137, this `Projection` is indeed eliminated in the `postOptimize()`. But the issue is that it affects the physical optimization and we have to eliminate it before physical optimization.

**Pull request merged**